cmd/tailscale/cli/serve.go

```diff
@@ -115,31 +229,457 @@ func (e *serveEnv) runServe(ctx context.Context, args []string) error {
+	}
+	hp := ipn.HostPort(net.JoinHostPort(dnsName, srvPortStr))
+	if !httpHandlerExists(sc, hp, mount) {
+		return errors.New("error: serve config does not exist")
+	}
+	// delete existing handler, then cascade delete if empty
+	delete(sc.Web[hp].Handlers, mount)
+	if len(sc.Web[hp].Handlers) == 0 {
+		delete(sc.Web, hp)
+		delete(sc.TCP, srvPort)
+	}
+	// clear empty maps mostly for testing
+	if len(sc.Web) == 0 {
+		sc.Web = nil
+	}
+	if len(sc.TCP) == 0 {
+		sc.TCP = nil
+	}
+	if err := e.setServeConfig(ctx, sc); err != nil {
+		return err
+	}
+	return nil
+}
+
+func httpHandlerExists(sc *ipn.ServeConfig, hp ipn.HostPort, mount string) bool {
+	_, ok := getHTTPHandler(sc, hp, mount)
+	return ok
+}
+
+func getHTTPHandler(sc *ipn.ServeConfig, hp ipn.HostPort, mount string) (handler *ipn.HTTPHandler, ok bool) {
+	if sc.Web != nil && sc.Web[hp] != nil &&
+		sc.Web[hp].Handlers != nil &&
+		sc.Web[hp].Handlers[mount] != nil {
+		return sc.Web[hp].Handlers[mount], true
+	}
+	return nil, false
+}
+
+func cleanMountPoint(mount string) (string, error) {
+	if mount == "" {
+		return "", errors.New("mount point cannot be empty")
+	}
+	if !strings.HasPrefix(mount, "/") {
+		mount = "/" + mount
+	}
+	c := path.Clean(mount)
+	if mount == c || mount == c+"/" {
+		return mount, nil
+	}
+	return "", fmt.Errorf("invalid mount point %q", mount)
+}
+
+func expandProxyTarget(target string) (string, error) {
+	if allNumeric(target) {
+		p, err := strconv.ParseUint(target, 10, 16)
+		if p == 0 || err != nil {
+			return "", fmt.Errorf("invalid port %q", target)
+		}
+		return "http://127.0.0.1:" + target, nil
+	}
+	if !strings.Contains(target, "://") {
+		target = "http://" + target
+	}
+	u, err := url.ParseRequestURI(target)
+	if err != nil {
+		return "", fmt.Errorf("parsing url: %w", err)
+	}
+	switch u.Scheme {
+	case "http", "https", "https+insecure":
+		// ok
+	default:
+		return "", fmt.Errorf("must be a URL starting with http://, https://, or https+insecure://")
+	}
+	host := u.Hostname()
+	switch host {
+	// TODO(shayne,bradfitz): do we want to do this?
+	case "localhost", "127.0.0.1":
+		host = "127.0.0.1"
+	default:
+		return "", fmt.Errorf("only localhost or 127.0.0.1 proxies are currently supported")
+	}
+	url := u.Scheme + "://" + host
+	if u.Port() != "" {
+		url += ":" + u.Port()
+	}
+	return url, nil
+}
+
+// isTCPForwardingAny checks if any TCP port is being forwarded.
+func isTCPForwardingAny(sc *ipn.ServeConfig) bool {
+	if sc == nil || len(sc.TCP) == 0 {
+		return false
+	}
+	for _, h := range sc.TCP {
+		if h.TCPForward != "" {
+			return true
+		}
+	}
+	return false
 }
 
-func (e *serveEnv) runServeShowConfig(ctx context.Context, args []string) error {
+// isTCPForwardingOnPort checks serve config to see if
+// we're specifically forwarding TCP on the given port.
+func isTCPForwardingOnPort(sc *ipn.ServeConfig, port uint16) bool {
+	if sc == nil || sc.TCP == nil || sc.TCP[port] == nil {
```

Comment:
it's safe to do reads on a nil map, so the sc.TCP == nil part is not needed

Reply:
Done.